File: verify_counts.py
import pandas as pd
from datetime import datetime, timedelta

file_path = "PAIE 02-2026 HORAIRE.xls"

try:
    # We will process both sheets just like PHP does
    sheets_to_process = ['HORAIRE', 'MENS', 'mens']
    excel_file = pd.ExcelFile(file_path)
    
    # Target dates
    dates = []
    # Nov 26 to Dec 25 mapped to 2026-01-26 to 2026-02-25
    for d in range(26, 32):
        dates.append(f"2026-01-{d:02d}")
    for d in range(1, 26):
        dates.append(f"2026-02-{d:02d}")
        
    all_blocks = []
    emp_count = 0
    
    for sheet_name in sheets_to_process:
        if sheet_name not in excel_file.sheet_names:
            continue
            
        df = pd.read_excel(file_path, sheet_name=sheet_name, header=1)
        
        # PHP logic checks rows starting from index 3 (which in this df would be index 1 because header is row 1 (0 in pandas for our header=1, wait...
        # In php, header: 1 returns array of arrays. row 0 is first row, row 1 is second row (dates). Data starts at row 3 (0,1,2,3 -> 4th row).
        # We can just iterate through the rows and look for employee numbers
        
        # The columns are 7 to 37 in pandas if it's 31 days
        date_cols = df.columns[6:37] # columns index 6 to 36
        
        for index, row in df.iterrows():
            matricule = str(row.iloc[0]).strip()
            if matricule == 'nan' or not matricule:
                continue
                
            emp_count += 1
            current_block = None
            emp_blocks = []
            
            for col_idx in range(6, 37): # 31 columns
                cell_val = str(row.iloc[col_idx]).strip().upper()
                date_str = dates[col_idx - 6]
                
                # Determine status exactly like PHP
                status = 'P'
                if cell_val == 'A':
                    status = 'A'
                elif cell_val == 'W' or '*' in cell_val:
                    status = 'W'
                elif cell_val in ('ML', 'M'):
                    status = 'M'
                elif cell_val in ('MT', 'MAT'):
                    status = 'MAT'
                elif cell_val in ('AT', 'ACC'):
                    status = 'ACC'
                else:
                    try:
                        float(cell_val)
                        status = 'P'
                    except ValueError:
                        status = 'P' # empty or unknown
                        if cell_val == 'NAN' or cell_val == '':
                            pass # Still considered a break in absence
                            
                # Grouping logic
                if status in ['A', 'M', 'MAT', 'ACC']:
                    if current_block and current_block['type'] == status:
                        # consecutive check
                        d1 = datetime.strptime(current_block['end_date'], "%Y-%m-%d")
                        d2 = datetime.strptime(date_str, "%Y-%m-%d")
                        diff = (d2 - d1).days
                        if diff <= 3:
                            current_block['end_date'] = date_str
                        else:
                            emp_blocks.append(current_block)
                            current_block = {'type': status, 'start': date_str, 'end_date': date_str}
                    else:
                        if current_block:
                            emp_blocks.append(current_block)
                        current_block = {'type': status, 'start': date_str, 'end_date': date_str}
                elif status == 'P' or cell_val == 'NAN' or cell_val == '':
                    if current_block:
                        emp_blocks.append(current_block)
                        current_block = None
            
            if current_block:
                emp_blocks.append(current_block)
                
            all_blocks.extend(emp_blocks)
            
    # Now filter by the report date range (Feb 1 to Feb 28)
    filter_start = datetime.strptime("2026-02-01", "%Y-%m-%d")
    filter_end = datetime.strptime("2026-02-28", "%Y-%m-%d")
    
    counts = {'A': 0, 'M': 0, 'MAT': 0, 'ACC': 0}
    
    for b in all_blocks:
        b_start = datetime.strptime(b['start'], "%Y-%m-%d")
        b_end = datetime.strptime(b['end_date'], "%Y-%m-%d")
        
        # Overlap condition
        if (filter_start <= b_start <= filter_end) or \
           (filter_start <= b_end <= filter_end) or \
           (b_start <= filter_start and b_end >= filter_end):
            if b['type'] in counts:
                counts[b['type']] += 1
                
    print(f"Total Employees Processed: {emp_count}")
    print(f"Total Absences Overlapping Feb 1 - Feb 28:")
    print(f"Maladie (M): {counts['M']}")
    print(f"Maternite (MAT): {counts['MAT']}")
    print(f"Accidents (ACC): {counts['ACC']}")
    print(f"Absences (A): {counts['A']}")
    print(f"Total: {sum(counts.values())}")
    
except Exception as e:
    print(f"Error: {e}")

File: verify_counts.js
const XLSX = require('xlsx');

const workbook = XLSX.readFile('PAIE 02-2026 HORAIRE.xls');
const sheets_to_process = ['HORAIRE', 'MENS', 'mens'];

const dates = [];
for (let d = 26; d <= 31; d++) dates.push(`2026-01-${String(d).padStart(2, '0')}`);
for (let d = 1; d <= 25; d++) dates.push(`2026-02-${String(d).padStart(2, '0')}`);

let all_blocks = [];
let emp_count = 0;

sheets_to_process.forEach(sheetName => {
    if (!workbook.Sheets[sheetName]) return;

    const sheet = workbook.Sheets[sheetName];
    const data = XLSX.utils.sheet_to_json(sheet, { header: 1, defval: null });

    // Data starts around row 3 (index 3 in 0-based if not strict header)
    for (let i = 3; i < data.length; i++) {
        const row = data[i];
        if (!row || row.length < 5) continue;

        let matricule = row[0];
        if (!matricule || String(matricule).trim() === '') continue;
        let fullname = row[1];
        if (!fullname || String(fullname).trim() === '') continue;

        emp_count++;
        let current_block = null;
        let emp_blocks = [];

        // columns 7 to 37 are the 31 days
        for (let col = 7; col <= 37; col++) {
            let cellValue = row[col];
            let dateStr = dates[col - 7];

            let cellStr = cellValue === null || cellValue === undefined ? '' : String(cellValue).trim().toUpperCase();

            let status = 'P';
            if (cellStr === 'A') {
                status = 'A';
            } else if (cellStr === 'W' || cellStr.includes('*')) {
                status = 'W';
            } else if (cellStr === 'ML' || cellStr === 'M') {
                status = 'M';
            } else if (cellStr === 'MT' || cellStr === 'MAT') {
                status = 'MAT';
            } else if (cellStr === 'AT' || cellStr === 'ACC') {
                status = 'ACC';
            } else {
                let parsed = parseFloat(cellStr);
                if (!isNaN(parsed)) {
                    status = 'P';
                } else {
                    continue; // Skip the cell completely, just like hr_import.js in PHP!
                }
            }

            if (['A', 'M', 'MAT', 'ACC'].includes(status)) {
                if (current_block && current_block.type === status) {
                    let d1 = new Date(current_block.end_date);
                    let d2 = new Date(dateStr);
                    let diffDays = (d2 - d1) / (1000 * 60 * 60 * 24);
                    if (diffDays <= 3) {
                        current_block.end_date = dateStr;
                    } else {
                        emp_blocks.push(current_block);
                        current_block = { type: status, start_date: dateStr, end_date: dateStr };
                    }
                } else {
                    if (current_block) emp_blocks.push(current_block);
                    current_block = { type: status, start_date: dateStr, end_date: dateStr };
                }
            } else if (status === 'P') {
                // Notice: In PHP, empty cells ('') are skipped by JS.
                // We only break if we see a genuine 'P' (number)
                if (current_block) {
                    emp_blocks.push(current_block);
                    current_block = null;
                }
            }
            // IF it's empty cellStr === '', we do nothing, meaning current_block remains open!
        }
        if (current_block) emp_blocks.push(current_block);
        
        all_blocks = all_blocks.concat(emp_blocks);
    }
});

const filter_start = new Date("2026-02-01");
const filter_end = new Date("2026-02-28");

const counts = { 'A': 0, 'M': 0, 'MAT': 0, 'ACC': 0 };

all_blocks.forEach(b => {
    let b_start = new Date(b.start_date);
    let b_end = new Date(b.end_date);

    if ((b_start >= filter_start && b_start <= filter_end) ||
        (b_end >= filter_start && b_end <= filter_end) ||
        (b_start <= filter_start && b_end >= filter_end)) {
        if (counts[b.type] !== undefined) {
            counts[b.type]++;
        }
    }
});

console.log(`Total Employees Processed: ${emp_count}`);
console.log(`\nExact counts of absences overlapping Feb 1st - Feb 28th based on Excel:`);
console.log(`Maternite (MAT): ${counts['MAT']}`);
console.log(`Maladie (M): ${counts['M']}`);
console.log(`Absences Injustifiees (A): ${counts['A']}`);
console.log(`Accidents de Travail (ACC): ${counts['ACC']}`);
console.log(`Total: ${counts['MAT'] + counts['M'] + counts['A'] + counts['ACC']}`);
